[FLINK-19082][doc] Add docs for temporal table and temporal table join (#14152)
apache/flink: docs/dev/table/streaming/joins.md

Compared to a regular join operation, this kind of join only supports append-only tables with time attributes. Since time attributes are quasi-monotonic increasing, Flink can remove old values from its state without affecting the correctness of the result.

Temporal Joins
--------------
<span class="label label-danger">Attention</span> This feature is only supported in Blink planner.

> <span class="label label-danger">Attention</span> Attention it is only supported in SQL, and not supported in Table API yet.

Reviewer: Aren't temporal joins still supported in the Table API using the old method (and listed [here](https://ci.apache.org/projects/flink/flink-docs-release-1.12/dev/table/tableApi.html#joins))? If so, this sentence is a bit misleading.

Author: I agree your point, @morsapaes . Old temporal joins(using temporal table function + `lateral`) is still supported in Table API, the newborn temporal join(using temporal table DDL + temporal join syntax) is supported in SQL.
How about :
```
Temporal table has ways to define in Flink, i.e. [temporal table function]({% link dev/table/streaming/temporal_tables.md %}#temporal-table-function) ) and [temporal table DDL](({% link dev/table/streaming/temporal_tables.md %}#temporal-table-ddl)).
Temporal join that using temporal table function is only supported in Table API,  temporal join that using temporal table DDL is only supported in SQL, please see the page [temporal table]({% link dev/table/streaming/temporal_tables.md %}) for more information about the differences between temporal table function and temporal table DDL.
```

Compared to a regular join operation, this kind of join only supports append-only tables with time attributes. Since time attributes are quasi-monotonic increasing, Flink can remove old values from its state without affecting the correctness of the result.

Temporal Joins
--------------

> <span class="label label-danger">Attention</span> This is only supported in Blink planner.

Reviewer 2: ```suggestion
<span class="label label-danger">Attention</span> This feature is only supported in Blink planner.
```

<span class="label label-danger">Attention</span> Attention It is only supported in SQL, and not supported in Table API yet.

Temporal join is an arbitrary table (left input/probe side) correlate with the versions of temporal table (right input/build side), The temporal table can be an external dimension table that changes over time 
or a changelog that tracks all history changes. Please check the corresponding page for more information about [temporal tables](temporal_tables.html).

> Flink uses the SQL syntax of `FOR SYSTEM_TIME AS OF` to query temporal table, which is proposed in SQL:2011. The syntax of temporal table join is as follows:

Reviewer 2: ```suggestion
Flink uses the SQL syntax of `FOR SYSTEM_TIME AS OF` to query temporal table, which is proposed in SQL:2011 standard. The syntax of a temporal table join is as follows:
```

{% highlight sql %}
SELECT [column_list]
FROM table1 [AS <alias1>]
[LEFT] JOIN table2 FOR SYSTEM_TIME AS OF table1.{ proctime | rowtime } [AS <alias2>]
ON table1.column-name1 = table2.column-name1
{% endhighlight %}

### Processing-time Temporal Joins

> Processing-time temporal join is the left input table using it's processing time to correlate the latest version of [regular table](temporal_tables.html#defining-regular-table) in a temporal join.

Reviewer 2: ```suggestion
Processing-time temporal join uses the left input tables processing time attribute to correlate the latest version of a [regular table](temporal_tables.html#defining-regular-table).
```

> Processing-time temporal join only supports regular table currently, and the supported regular table table can only contain append-only stream. 

Reviewer 2: ```suggestion
Processing-time temporal join only supports regular tables currently, and the supported regular table can only contain append-only stream. 
```

> With a processing-time time attribute, it is impossible to pass _past_ time attributes as an argument to the temporal table.

Reviewer 2: ```suggestion
With a processing-time time attribute, it is impossible to use a _past_ time attribute as an argument to the temporal table.
```

For example, we would like to convert the following orders using the latest rate in `LatestRates`. The result would be:

{% highlight text %}
amount currency     rate   amout*rate
====== ========= ======= ============
     2 Euro          114          228    <== arrived at time 10:15
     1 US Dollar     102          102    <== arrived at time 10:30
     2 Euro          116          232    <== arrived at time 10:52
{% endhighlight %}

With the help of temporal table join, we can express such a query in SQL as:

{% highlight sql %}
SELECT
  o.amout, o.currency, r.rate, o.amount * r.rate
FROM
  Orders AS o
  JOIN LatestRates FOR SYSTEM_TIME AS OF o.proctime AS r
  ON r.currency = o.currency
{% endhighlight %}

Each record from the probe side will be joined with the current version of the build side table. In our example, the query is using the processing-time notion, so a newly appended order would always be joined with the most recent version of `LatestRates` when executing the operation.
Note that the result is not deterministic for processing-time.
The processing-time temporal join is usually used to enrich the stream with external table (i.e. dimension table).

> ### Event-time Temporal Joins

Reviewer 2: Can you please move event time temporal joins above processing time? I think this is the more common use case.

Event-time temporal join is the left input table using it's event time to correlate the corresponding version of [versioned table](temporal_tables.html#defining-versioned-table) in a temporal join.

> Event-time temporal join only supports versioned table, the versioned table table need to be a changelog stream, the append-only stream can convert to changelog stream in Flink, thus the versioned table can come from an append-only stream,

Reviewer 2: ```suggestion
Event-time temporal join only supports versioned tables and the versioned tables need to be a changelog stream. However, an append-only stream can be converted to a changelog stream in Flink, thus the versioned table can come from an append-only stream,
```

please see [versioned view](temporal_tables.html#defining-versioned-view) for more information to know how to define a versioned table from an append-only stream.

> With an event-time time attribute (i.e., a rowtime attribute), it is possible to pass _past_ time attributes to the temporal table.

Reviewer 2: ```suggestion
With an event-time time attribute (i.e., a rowtime attribute), it is possible to use a _past_ time attribute with the temporal table.
```
